packages/timeout/src/index.ts

```diff
@@ -0,0 +1,111 @@
+// eslint-disable-next-line max-classes-per-file
+import { setTimeout } from 'timers/promises';
+import { AsyncLocalStorage } from 'async_hooks';
+
+export interface AbortContext {
+    cancelTimeout: AbortController;
+    cancelTask: AbortController;
+}
+
+/**
+ * `AsyncLocalStorage` instance that is used for baring the AbortContext inside user provided handler.
+ * We can use it to access the `AbortContext` instance via `storage.getStore()`, and there we can access
+ * both `cancelTimeout` and `cancelTask` instances of `AbortController`.
+ */
+export const storage = new AsyncLocalStorage<AbortContext>();
+
+/**
+ * Custom error class that will be used for timeout error.
+ */
+export class TimeoutError extends Error {
+}
+
+/**
+ * Custom error class to handle `tryCancel()` checks.
+ * This should not be exposed to user land, as it will be caught in.
+ */
+class InternalTimeoutError extends TimeoutError {
+}
+
+/**
+ * Checks whether we are inside timeout handler created by this package, and cancels current
+ * task execution by throwing `TimeoutError`. This error will be ignored if the promise timed
+ * out already, or explicitly skipped in `addTimeoutToPromise`.
+ *
+ * Use this function after every async call that runs inside the timeout handler:
+ *
+ * ```ts
+ * async function doSomething() {
+ *     await doSomethingTimeConsuming();
+ *     tryCancel();
+ *     await doSomethingElse();
+ *     tryCancel();
+ * }
+ * ```
+ */
+export function tryCancel(): void {
+    const { signal } = storage.getStore()?.cancelTask ?? {};
+
+    if (signal?.aborted) {
+        throw new InternalTimeoutError('Promise handler has been canceled due to a timeout');
+    }
+}
+
+/**
+ * Runs given handler and rejects with the given `errorMessage` (or `Error` instance)
+ * after given `timeoutMillis`, unless the original promise resolves or rejects earlier.
+ * Use `tryCancel()` function inside the handler after each await to finish its execution
+ * early when the timeout appears.
+ *
+ * ```ts
+ * const res = await addTimeoutToPromise(
+ *   () => handler(),
+ *   200,
+ *   'Handler timed out after 200ms!',
+ * );
+ * ```
+ */
+export async function addTimeoutToPromise<T>(handler: () => Promise<T>, timeoutMillis: number, errorMessage: string | Error): Promise<T> {
+    // respect existing context to support nesting
+    const context = storage.getStore() ?? {
+        cancelTimeout: new AbortController(),
+        cancelTask: new AbortController(),
+    };
+    let ret: T;
```

Comment:
Would it hurt to name it `returnValue` ? 😄

Reply:
Probably yes 😁